### PR TITLE
RDM-12680 - Search Criteria

### PR DESCRIPTION
**Pull request by KleoG**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-12680

### Change description ###
Update the Definition Store to recognise the New SearchCriteria Tab

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: repository/src/test/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            () -> assertThat(result, hasItem(hasProperty("otherCaseReference", is(OTHER_CASE_REFERENCE))))
+        );
+    }
+
+    @Test
+    public void shouldReturnNoSearchCriteriaResultForUnknownParameters() {
+        List<SearchCriteriaEntity> result = searchCriteriaRepository
+            .findSearchCriteriaEntityByCaseType("IncorrectCaseType");
+
+        assertAll(
```
Comment (sullivangareth) [inline repository/src/test/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepositoryTest.java:68]: Dont need assertAll here ?

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/EntityToResponseDTOMapperTest.java [hunk 2]
```diff
@@ -1764,6 +1766,47 @@ private CaseTypeEntity caseTypeEntity(String reference) {
...
+            SearchCriteria searchCriteria = classUnderTest.map(searchCriteriaEntity);
+
+            assertNull(searchCriteria.getCaseTypeId());
+            assertNull(searchCriteria.getLiveFrom());
+            assertNull(searchCriteria.getLiveTo());
+            assertNull(searchCriteria.getOtherCaseReference());
+        }
+
+
+        private CaseTypeEntity caseTypeEntity(String reference) {
```
Comment (sullivangareth) [inline domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/EntityToResponseDTOMapperTest.java:1802]: This code is replicated in several of the nested classes within this class - could it be defined once in top level class and reused ? Appreciate its a pattern thats been copied ...

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import static org.mockito.ArgumentMatchers.anyList;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+class SearchCriteriaServiceImplTest {
+
+    public static final String OTHER_CASE_REFERENCE = "OtherCaseReference";
```
Comment (sullivangareth) [inline domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java:29]: private ?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/SeaarchCriteriaParserTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+class SeaarchCriteriaParserTest extends ParserTestBase {
+
+    private static final String CASE_TYPE_ID_1 = "TestCaseTypeID_1";
+    private static final String CASE_TYPE_ID_2 = "TestCaseTypeID_2";
+
+    public static final String OTHER_CASE_REFERENCE = "OtherCaseReference";
```
Comment (sullivangareth) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/SeaarchCriteriaParserTest.java:30]: private ?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchCriteriaValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        // THEN
+        verifyDotNotationValidatorCallNeverMadeFor(ColumnName.OTHER_CASE_REFERENCE);
+    }
+
+    @DisplayName("throws exception if OtherCaseReference validation fails")
+    @Test
+    void throwsExceptionIfOtherCaseReferenceValidationFails() {
+
+        // GIVEN
+        SearchCriteriaEntity searchCriteriaEntity1 = createBlankSearchCriteriaEntity();
```
Comment (sullivangareth) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchCriteriaValidatorTest.java:103]: remove `1` from name ?

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaService.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import uk.gov.hmcts.ccd.definition.store.repository.entity.SearchCriteriaEntity;
+import uk.gov.hmcts.ccd.definition.store.repository.model.SearchCriteria;
+
+import java.util.List;
+
+public interface SearchCriteriaService {
+
+    void saveAll(List<SearchCriteriaEntity> entityList);
+
+    List<SearchCriteria> findByCaseTypeReferences(List<String> caseTypeReferences);
```
Comment (smathangi) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaService.java:12]: where do we use this method? `findByCaseTypeReferences` ?

Comment (smathangi) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaService.java:12]: If not used can you remove from SearchPartyService as well?

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        doReturn(searchCriteriaEntities).when(repository).findByCaseTypeReferenceIn(anyList());
+        List<String> references = Lists.newArrayList("Test", "Divorce");
+        List<SearchCriteria> valuesReturned = sut.findByCaseTypeReferences(references);
+        Assert.assertEquals(2, valuesReturned.size());
+    }
+
+    @Test
+    @DisplayName(
+        "Should get empty Role To Access Profiles list for the passed empty case type references")
+    void shouldGetEmptyRoleToAccessProfilesListForEmptyCaseType() {
```
Comment (smathangi) [inline domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java:64]: test name & description are not reflecting calls..
please see https://github.com/hmcts/ccd-definition-store-api/pull/1061/files#r692719089

File: repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepository.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.data.repository.query.Param;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.SearchCriteriaEntity;
+
+import java.util.List;
+
+import static uk.gov.hmcts.ccd.definition.store.repository.QueryConstants.SELECT_LATEST_CASE_TYPE_ENTITY_FOR_REFERENCE;
+
+public interface SearchCriteriaRepository extends JpaRepository<SearchCriteriaEntity, Integer> {
+
+    List<SearchCriteriaEntity> findByCaseTypeReferenceIn(List<String> caseTypeReferences);
```
Comment (smathangi) [inline repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepository.java:14]: same as like this https://github.com/hmcts/ccd-definition-store-api/pull/1061/files#r692719089

File: repository/src/test/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepositoryTest.java (hunk 1 above)
Comment (KleoG) [reply repository/src/test/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepositoryTest.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/EntityToResponseDTOMapperTest.java (hunk 2 above)
Comment (KleoG) [reply domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/EntityToResponseDTOMapperTest.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java (hunk 3 above)
Comment (KleoG) [reply domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/SeaarchCriteriaParserTest.java (hunk 4 above)
Comment (KleoG) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/SeaarchCriteriaParserTest.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchCriteriaValidatorTest.java (hunk 5 above)
Comment (KleoG) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchCriteriaValidatorTest.java:103]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaService.java (hunk 6 above)
Comment (KleoG) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaService.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java (hunk 7 above)
Comment (KleoG) [reply domain/src/test/java/uk/gov/hmcts/ccd/definition/store/domain/service/searchcriteria/SearchCriteriaServiceImplTest.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

File: repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepository.java (hunk 8 above)
Comment (KleoG) [reply repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/SearchCriteriaRepository.java]: done in [59c375d1](https://github.com/hmcts/ccd-definition-store-api/pull/1061/commits/59c375d172d81b375f2fe43039955b022afd906e)

**Review by smathangi — APPROVED** [review]
LGTM

**Pull request merged**